### PR TITLE
[TE] Auto-Resolving Create-Update API for alerts

### PR DESCRIPTION
**Pull request by akshayrai**
A single endpoint to create/update alert based on the detectionName provided in the yaml config.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 1]
```diff
@@ -414,6 +416,70 @@ public Response updateDetectionPipelineApi(
...
+    return existingDetectionConfig;
+  }
+
+  long createUpdateDetectionPipeline(String payload) {
+    Preconditions.checkArgument(StringUtils.isNotBlank(payload), "The Yaml Payload in the request is empty.");
+    long detectionId;
+    DetectionConfigDTO existingDetection = fetchExistingDetection(payload);
+    if (existingDetection != null) {
+      detectionId = existingDetection.getId();
+      updateDetectionPipeline(detectionId, payload, 0, 0);
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: Let's overload updateDetectionPipeline with only first two params.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 2]
```diff
@@ -414,6 +416,70 @@ public Response updateDetectionPipelineApi(
...
+    Collection<DetectionConfigDTO> detectionConfigs = this.detectionConfigDAO
+        .findByPredicate(Predicate.EQ("name", detectionName));
+    if (detectionConfigs != null && !detectionConfigs.isEmpty()) {
+      existingDetectionConfig = detectionConfigs.iterator().next();
+    }
+
+    return existingDetectionConfig;
+  }
+
+  long createUpdateDetectionPipeline(String payload) {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: Use name "createOrUpdateDetectionPipeline"

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 3]
```diff
@@ -414,6 +416,70 @@ public Response updateDetectionPipelineApi(
...
+  /**
+   Set up a detection pipeline using a YAML config - create new or update existing
+   @param payload YAML config string
+   @return a message contains the saved detection config id
+   */
+  @POST
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.TEXT_PLAIN)
+  @ApiOperation("Create a new detection pipeline or update existing if one already exists")
+  public Response createUpdateDetectionPipelineApi(@ApiParam("yaml config") String payload) {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: use "createOrUpdatePipelineApi"

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 4]
```diff
@@ -414,6 +416,70 @@ public Response updateDetectionPipelineApi(
...
+    Map<String, String> responseMessage = new HashMap<>();
+    long detectionConfigId;
+    try {
+      detectionConfigId = createUpdateDetectionPipeline(payload);
+    } catch (IllegalArgumentException e) {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: The catch blocks are duplicated in YamlResource. Suggest to put them together.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java (hunk 1 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: Addressed

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java (hunk 2 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: Updated

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java (hunk 3 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: Updated

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java (hunk 4 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: @xiaohui-sun, I didn't quite get the point. Can you provide some details?

Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: There are multiple places to handle pipeline creation exceptions. If you search for "Validation error while creating subscription group with payload" there are already 2 matches. This is a minor  .

Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: I have addressed it in a separate PR: https://github.com/apache/incubator-pinot/commit/1ff23c3628e2148910da940d1a84afb791a9a3d9

**Pull request merged**